state_synchronizer/src/peer_manager.rs

```diff
@@ -68,12 +82,33 @@ impl PeerManager {
         self.get_active_upstream_peer_ids().is_empty()
     }
 
+    pub fn update_score(&mut self, peer_id: &PeerId, update_type: PeerScoreUpdateType) {
+        if let Some(peer_info) = self.peers.get_mut(peer_id) {
+            match update_type {
+                PeerScoreUpdateType::Success => {
+                    if peer_info.score < 100 {
+                        peer_info.score += 1;
+                    }
+                }
+                PeerScoreUpdateType::InvalidChunk => {
+                    peer_info.score -= 1;
+                }
+            }
+        }
+    }
+
     pub fn pick_peer(&mut self) -> Option<(PeerId, StateSynchronizerSender)> {
         let active_peers = self.get_active_upstream_peer_ids();
         debug!("[state sync] (pick_peer) state: {:?}", self.peers);
         if !active_peers.is_empty() {
-            let idx = thread_rng().gen_range(0, active_peers.len());
-            let peer_id = *active_peers[idx];
+            let weights: Vec<_> = active_peers
+                .iter()
+                .map(|peer_id| self.peers.get(peer_id).unwrap().score)
+                .collect();
+            let dist = WeightedIndex::new(&weights).unwrap();
```

Comment:
ideally we don't have to reconstruct WeightedIndex on each call if there's no change in weights. But we can optimize it later

Reply:
I'll try optimizing in the following diff, probably just store the weighted index as a param in peer manager and update whenever score is updated